Makefile.gcc4

```diff
@@ -0,0 +1,33 @@
+# Makefile for ancient gcc4 (tested on Ubuntu 7.10 with 4.1.2-16ubuntu2)
+# Compiles Kaitai Struct C++ runtime and tests
+
+BUILD_DIR := build
+
+SOURCES := \
+	kaitai/kaitaistream.cpp \
+	tests/unittest.cpp
+
+OBJS := \
+	$(BUILD_DIR)/kaitaistream.o \
+	$(BUILD_DIR)/unittest.o
+
+DEFINES := -DKS_STR_ENCODING_ICONV -DGTEST_NANO
+
+LDFLAGS :=
+LDLIBS :=
+
+INCLUDES := -I.
+
+all: $(BUILD_DIR) $(BUILD_DIR)/unittest
+
+$(BUILD_DIR):
+	mkdir -p $(BUILD_DIR)
```

Comment:
Actually, I'm a little bit unsure whether this "directory rule" will work correctly in `make` (and whether `make` will understand that this `$(BUILD_DIR)` rule in fact is a prerequisite to of all the `$(BUILD_DIR)/...` stuff). I wonder whether it shouldn't be added as a dependency of the `$(BUILD_DIR)/...` rules in some way.

It seems that people on the internet do the same stuff as we do here, and [in this SO answer](https://stackoverflow.com/a/4251368/12940655) someone suggests that GNU make has a concept of [order-only prerequisites](http://www.gnu.org/s/hello/manual/make/Prerequisite-Types.html) that are supposedly appropriate for this case, but I've never heard of it.

To me, the easiest workaround that avoids all these questions would be to have the `build/` directory tracked by Git again - I see you removed it from Git index in https://github.com/kaitai-io/kaitai_struct_cpp_stl_runtime/commit/aa6e59b556b451674964e2ecf4a88ec73b2b454e, which I personally have no problem with, but it would've avoided these problems.

Reply:
Sharing my point of view on this: in general, this whole Makefile-something is a hack, which exists only to prove certain degree of compatibility with ancient compilers. It doesn't have to be particularly robust or be well-tested in all possible scenarios.

I completely understand that Makefiles per se are a pretty dangerous conversation to have. You start with something simple (2-3 two-liner rules), and then you start generalizing, you start supporting all possible cleanups and permutations of partial builds, and several weeks after you realize that your Makefile is incomprehensible mess of 1000s lines of code with lots of obscure syntax and ifdefs. And we haven't even started touching the topic of GNU Make vs BSD Make, etc.

This is pretty much exactly the reason why vast majority of even 1990s C/C++ projects strive away from Makefiles altogether, at the very least abstracting this part out with autotools or cmake or anything else.

In this specific case, I would leave it as is, as long as it moderately good does its only job: proving that KS C++ runtime works on Ubuntu 7.10.

Having directories like `build/` added to version control is generally a bad idea — if anything, for sole reason of being able to deleting whole `build/` to reset and not worry about version control crying about lost gitignore and/or placeholder file. If a build process generates `build/tests/Debug/...`, why shouldn't it create `build/` in the first place?

Reply:
> Sharing my point of view on this: in general, this whole Makefile-something is a hack, which exists only to prove certain degree of compatibility with ancient compilers. It doesn't have to be particularly robust or be well-tested in all possible scenarios.

Fair enough, I'm completely OK with this. Probably I tend to overgeneralize things, but I see the benefit of writing code that is safe to use not only in the one place for which it was intended, but also safe for inspiration when there's a need to do a similar thing for another purpose in the future, without having to analyze the differences of situations and why it may have been OK to use a pattern in one place but not in this other place.

Reply:
Totally get your aspiration, but for this specific case I would say that we should rather discourage people from building on top of ancient compilers :)